src/v/kafka/server/tests/produce_partition_bench.cc

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "kafka/client/types.h"
+#include "kafka/protocol/fetch.h"
+#include "kafka/protocol/schemata/produce_request.h"
+#include "kafka/protocol/types.h"
+#include "kafka/server/handlers/produce.h"
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "random/generators.h"
+#include "redpanda/tests/fixture.h"
+#include "test_utils/fixture.h"
+
+#include <seastar/core/sstring.hh>
+#include <seastar/testing/perf_tests.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <boost/range/iterator_range_core.hpp>
+#include <boost/test/tools/interface.hpp>
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test_log.hpp>
+#include <fmt/ostream.h>
+
+#include <tuple>
+
+using namespace std::chrono_literals; // NOLINT
+
+struct produce_partition_fixture : redpanda_thread_fixture {
+    static constexpr size_t topic_name_length = 30;
+    static constexpr size_t total_partition_count = 1;
+
+    model::topic t;
+
+    produce_partition_fixture() {
+        BOOST_TEST_CHECKPOINT("before leadership");
+
+        wait_for_controller_leadership().get0();
+
+        BOOST_TEST_CHECKPOINT("HERE");
+
+        t = model::topic(
+          random_generators::gen_alphanum_string(topic_name_length));
+        auto tp = model::topic_partition(t, model::partition_id(0));
+        add_topic(
+          model::topic_namespace_view(model::kafka_namespace, t),
+          total_partition_count)
+          .get();
+
+        auto ntp = make_default_ntp(tp.topic, tp.partition);
+        wait_for_leader(ntp).get0();
+        BOOST_TEST_CHECKPOINT("HERE");
+    }
+};
+
+PERF_TEST_C(produce_partition_fixture, test_produce_partition) {
+    BOOST_TEST_CHECKPOINT("HERE");
+
+    model::topic_partition tp = model::topic_partition(
+      t, model::partition_id(0));
+
+    storage::record_batch_builder builder(
+      model::record_batch_type::raft_data, model::offset{0});
+
+    constexpr size_t num_records = 100;
+    for (size_t i = 0; i < num_records; ++i) {
+        builder.add_raw_kv(iobuf{}, iobuf{});
```

Comment:
Can we put some data in the messages? Seems like something easy to make the test parameterizable on: the size of the payload.